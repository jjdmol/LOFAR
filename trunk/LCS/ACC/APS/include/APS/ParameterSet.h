//#  ParameterSet.h: ParameterCollectin filled with runtime values.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  Abstract:
//#
//#	 Defines a class the should contain fully filled ParamaterCollections
//#  to be used in runtime to feed the applications with information.
//#
//#  $Id$

#ifndef ACC_PARAMETERSET_H
#define ACC_PARAMETERSET_H

#include <lofar_config.h>

//# Includes
#include <ACC/ParameterCollection.h>

namespace LOFAR {
  namespace ACC {


//# Description of the class.
// The ParameterSet class is a ParameterCollection that is used during runtime
// to feed an application with its runtime values.
// The restrictions of this collections are:
// 1. The firstline must be a versionnr key with a valid versionnumber.
// 2. All values must be filled in.
//
class ParameterSet : public ParameterCollection
{
public:
	// Default constructable;
	ParameterSet();
	~ParameterSet();

	// Define a conversion function from base class to this class
	ParameterSet(const ParameterCollection& that);

	// Allow reading a file for backwards compatibility
	explicit ParameterSet(const string&	theFilename);

	// Copying is allowed.
	ParameterSet(const ParameterSet& that);
	ParameterSet& 	operator=(const ParameterSet& that);

	// Check if the contents is a valid ParameterSet.
	bool check(string&	errorReport) const;

	friend std::ostream& operator<<(std::ostream& os, const ParameterCollection &thePS);

private:
};

} // namespace ACC
} // namespace LOFAR

#endif
