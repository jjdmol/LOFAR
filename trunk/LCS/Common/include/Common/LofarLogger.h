//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#if !defined(COMMON_LOFAR_LOGGER_H)
#define COMMON_LOFAR_LOGGER_H

#include <lofar_config.h>
#include <Common/StringUtil.h>

//# Dependent of the availability of the log4Cplus package one set of macro's
//# or the other set of macro's is included.
#ifdef HAVE_LOG4CPLUS
#include <Common/LofarLog4Cplus.h>
#else
#include <Common/LofarLogCout.h>
#endif //HAVE_LOG4CPLUS

namespace LOFAR 
{
  // Define the LOFAR::Exception class.
  EXCEPTION_CLASS(AssertError,Exception)  
}

//# --------------------- Common definitions ---------------

//# Make sure that LOFARLOGGER_PACKAGE is defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

//#
//# AUTO_FUNCTION_NAME
//#
// The macro AUTO_FUNCTION_NAME will be resolved by the (pre)compiler to 
// hold the name of the function the macro was used in.
#if defined(HAVE_PRETTY_FUNCTION)
#	define AUTO_FUNCTION_NAME		__PRETTY_FUNCTION__
#elif defined(HAVE_FUNCTION)
#	define AUTO_FUNCTION_NAME		__FUNCTION__
#else
#	define AUTO_FUNCTION_NAME		"??"
#endif

// NB: THROW is defined in package-dependant file.

// If the condition of assert is NOT met a logrequest is sent to the logger
// <module>.EXCEPTION and an AsserError exception is thrown.
#define ASSERT(cond) \
	if (!(cond))  THROW(LOFAR::AssertError, "Assertion: " #cond)
#define ASSERTSTR(cond,stream) \
	if (!(cond))  THROW(LOFAR::AssertError, "Assertion: " #cond "; " << stream)

// If the condition of failwhen is met a logrequest is sent to the logger
// <module>.EXCEPTION and an AsserError exception is thrown.
#define FAILWHEN(cond) \
	if (cond)  THROW(LOFAR::AssertError, "Failtest: " #cond)

// The DBG... version of ASSERT and FAILWHEN will only be in your compiled
// code when the (pre)compiler flag ENABLE_DBGASSERT is defined.
#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

#endif // file read before
