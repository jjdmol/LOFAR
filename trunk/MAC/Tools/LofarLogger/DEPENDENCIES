Compiler version:
compat-2.96 (special compiler for the RedHat 8.0 distribution; almost the same as 2.95 but extended with STL)
Because the 3.2 is already installed with RedHat 8.0 the gcc and g++ of the compat 
installation must be renamed (to for instance gcc296 and g++296).

EXTERNAL PACKAGES:
		
log4cplus-1.0
	Prepare build:
	> CC=gcc296 CXX=g++296 CXXFLAGS="-g -O2 -fno-rtti" sh ./configure --enable-threads=no
	- set '#define HAVE_SSTREAM 1' to comment
	- ignore compiler erros during compiling the 'test''s
	- ignore compiler error during installing the 'test's

LCS/Common
