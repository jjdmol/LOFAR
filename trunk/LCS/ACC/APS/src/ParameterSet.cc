//#  ParameterSet.cc: Implements a map of Key-Value pairs.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>


#include <APS/ParameterSet.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_fstream.h>


namespace LOFAR {
  namespace ACC {
    namespace APS {

//-------------------------- creation and destroy ---------------------------
static ParameterSet* globalParameterSetInstance = 0;

ParameterSet* globalParameterSet()
{
  if (globalParameterSetInstance == 0) {
    globalParameterSetInstance = new ParameterSet();
  }
  return (globalParameterSetInstance);
}

//-------------------------- creation and destroy ---------------------------
//
// Default constructor
//
ParameterSet::ParameterSet(KeyCompare::Mode	mode)
  : itsSet (new ParameterSetImpl(mode))
{}

ParameterSet::ParameterSet(bool caseInsensitive)
  : itsSet (new ParameterSetImpl(caseInsensitive ?
                                 KeyCompare::NOCASE : KeyCompare::NORMAL))
{}

//
// Construction by reading a parameter file.
//
ParameterSet::ParameterSet(const string& theFilename, bool caseInsensitive)
  : itsSet (new ParameterSetImpl(theFilename, caseInsensitive ?
                                 KeyCompare::NOCASE : KeyCompare::NORMAL))
{}

//
// Construction by reading a parameter file.
//
ParameterSet::ParameterSet(const string&	theFilename,
			   KeyCompare::Mode	mode)
  : itsSet (new ParameterSetImpl(theFilename, mode))
{}

ParameterSet::ParameterSet(const char*	theFilename,
			   KeyCompare::Mode	mode)
  : itsSet (new ParameterSetImpl(string(theFilename), mode))
{}

//
// Copying is allowed.
//
ParameterSet::ParameterSet(const ParameterSet& that)
  : itsSet (that.itsSet->incrCount())
{}

//
// operator= copying
//
ParameterSet& 
ParameterSet::operator=(const ParameterSet& that)
{
	if (this != &that) {
	  unlink();
	  itsSet = that.itsSet->incrCount();
	}
	return (*this);
}

//
//	Destructor
//
ParameterSet::~ParameterSet()
{
  unlink();
}

void ParameterSet::unlink()
{
  if (itsSet->decrCount() == 0) {
    delete itsSet;
  }
}

vector<int16> ParameterSet::getInt16Vector(const string& aKey,
                                           bool expandable) const
{
  vector<int> vec (getIntVector(aKey, expandable));
  return vector<int16> (vec.begin(), vec.end());
}

vector<int16> ParameterSet::getInt16Vector(const string& aKey,
                                           const vector<int16>& aValue,
                                           bool expandable) const
{
  vector<int> vec (getIntVector(aKey,
                                vector<int>(aValue.begin(), aValue.end()),
                                expandable));
  return vector<int16> (vec.begin(), vec.end());
}

vector<uint16> ParameterSet::getUint16Vector(const string& aKey,
                                             bool expandable) const
{
  vector<uint> vec (getUintVector(aKey, expandable));
  return vector<uint16> (vec.begin(), vec.end());
}

vector<uint16> ParameterSet::getUint16Vector(const string& aKey,
                                             const vector<uint16>& aValue,
                                             bool expandable) const
{
  vector<uint> vec (getUintVector(aKey,
                                  vector<uint>(aValue.begin(), aValue.end()),
                                  expandable));
  return vector<uint16> (vec.begin(), vec.end());
}

vector<int32> ParameterSet::getInt32Vector(const string& aKey,
                                           bool expandable) const
{
  vector<int> vec (getIntVector(aKey, expandable));
  return vector<int32> (vec.begin(), vec.end());
}

vector<int32> ParameterSet::getInt32Vector(const string& aKey,
                                           const vector<int32>& aValue,
                                           bool expandable) const
{
  vector<int> vec (getIntVector(aKey,
                                vector<int>(aValue.begin(), aValue.end()),
                                expandable));
  return vector<int32> (vec.begin(), vec.end());
}

vector<uint32> ParameterSet::getUint32Vector(const string& aKey,
                                             bool expandable) const
{
  vector<uint> vec (getUintVector(aKey, expandable));
  return vector<uint32> (vec.begin(), vec.end());
}

vector<uint32> ParameterSet::getUint32Vector(const string& aKey,
                                             const vector<uint32>& aValue,
                                             bool expandable) const
{
  vector<uint> vec (getUintVector(aKey,
                                  vector<uint>(aValue.begin(), aValue.end()),
                                  expandable));
  return vector<uint32> (vec.begin(), vec.end());
}


//
// operator<<
//
std::ostream&	operator<< (std::ostream& os, const ParameterSet &thePS)
{
	os << *thePS.itsSet;
	return os;
}

    } // namespace APS
  } // namespace ACC
} // namespace LOFAR
