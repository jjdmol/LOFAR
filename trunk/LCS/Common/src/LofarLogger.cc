//#  LofarLogger.cc: Interface to the log4cplus logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Includes
#include <Common/LofarLogger.h>

namespace LOFAR {

//#------------------------- Global user functions -------------------------------
//
// formatString(format, ...) --> string
//
// Define a global function the accepts printf like arguments and returns a string.
//
const std::string formatString(const	char* format, ...) {
	char		tmp_cstring[1024];
	va_list		ap;

	va_start (ap, format);
	vsnprintf(tmp_cstring, sizeof(tmp_cstring), format, ap);
	va_end   (ap);

	return   std::string(tmp_cstring);
}


//#------------------------- Internal implementation -----------------------------
using namespace log4cplus;

//# ------------------------ implement the five trace levels ------------------------
const LogLevel	TRACE1_LOG_LEVEL	= 1;
const LogLevel	TRACE2_LOG_LEVEL	= 2;
const LogLevel	TRACE3_LOG_LEVEL	= 3;
const LogLevel	TRACE4_LOG_LEVEL	= 4;
const LogLevel	TRACE5_LOG_LEVEL	= 5;

#define _TRACE1_STRING	LOG4CPLUS_TEXT("TRACE1")
#define _TRACE2_STRING	LOG4CPLUS_TEXT("TRACE2")
#define _TRACE3_STRING	LOG4CPLUS_TEXT("TRACE3")
#define _TRACE4_STRING	LOG4CPLUS_TEXT("TRACE4")
#define _TRACE5_STRING	LOG4CPLUS_TEXT("TRACE5")

tstring traceLevel2String(LogLevel ll) {
	switch (ll) {
	case TRACE1_LOG_LEVEL:	return _TRACE1_STRING;
	case TRACE2_LOG_LEVEL:	return _TRACE2_STRING;
	case TRACE3_LOG_LEVEL:	return _TRACE3_STRING;
	case TRACE4_LOG_LEVEL:	return _TRACE4_STRING;
	case TRACE5_LOG_LEVEL:	return _TRACE5_STRING;
	}

	return tstring();		// not found
}

LogLevel string2TraceLevel (const tstring& lname) {
	if (lname == _TRACE1_STRING)	return TRACE1_LOG_LEVEL;
	if (lname == _TRACE2_STRING)	return TRACE2_LOG_LEVEL;
	if (lname == _TRACE3_STRING)	return TRACE3_LOG_LEVEL;
	if (lname == _TRACE4_STRING)	return TRACE4_LOG_LEVEL;
	if (lname == _TRACE5_STRING)	return TRACE5_LOG_LEVEL;

	return NOT_SET_LOG_LEVEL;			// not found
}

// initTracemodule
//
// Function that is used when the TRACE levels are NOT compiled out. It registers
// the TRACEn management routines at the Log4Cplus LogLevelManager and sets up the
// global trace-logger named "TRACE", with the additivity set to false.
// Attached to the trace-logger is one Appender that logs to stderr.
//
void initTraceModule (void) {
	//# register our own loglevels
	getLogLevelManager().pushToStringMethod(traceLevel2String);
	getLogLevelManager().pushFromStringMethod(string2TraceLevel);
	
	//# Setup a property object to initialise the TRACE Logger
	helpers::Properties		traceProp;		
	traceProp.setProperty("log4cplus.logger.TRACE", "TRACE, TRACE");
	traceProp.setProperty("log4cplus.additivity.TRACE", "false");
	traceProp.setProperty("log4cplus.appender.TRACE", "log4cplus::ConsoleAppender");
//	traceProp.setProperty("log4cplus.appender.TRACE.layout","log4cplus::PatternLayout");
//	traceProp.setProperty("log4cplus.appender.TRACE.layout.ConversionPattern","%c %m%n");
	traceProp.setProperty("log4cplus.appender.TRACE.layout", "log4cplus::TTCCLayout");
	traceProp.setProperty("log4cplus.appender.TRACE.layout.Use_gmtime", "true");

	PropertyConfigurator(traceProp).configure();

}
LOFAR::LoggerReference	TraceLoggerRef("TRACE");			// create the tracelogger

}	// namespace LOFAR
