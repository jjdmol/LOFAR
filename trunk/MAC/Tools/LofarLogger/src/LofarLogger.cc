//#  LofarLogger.cc: logger wrapper
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#include "LofarLogger.h"
#include <stdarg.h>
#include <stdio.h>
#include <log4cplus/configurator.h>
#include <log4cplus/loggingmacros.h>

// namespaces
using namespace std;

// static member initialization
LofarLogger LofarLogger::_sLofarLogger;

/**
 * LofarLogger constructor
 */
LofarLogger::LofarLogger() :
    _loggerMap()
{
    log4cplus::PropertyConfigurator::doConfigure("log4cplus.properties");
}

LofarLogger::LofarLogger(const LofarLogger& rhs) :
    _loggerMap(rhs._loggerMap)
{
}

/**
 * ~LofarLogger destructor
 */
LofarLogger::~LofarLogger()
{
}

/**
 * getInstance - returns the static instance of the LofarLogger
 */
LofarLogger LofarLogger::getInstance()
{
  return _sLofarLogger;
}

/**
 * trace - logs TRACE level messages
 */
void LofarLogger::trace(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::TRACE_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::TRACE_LOG_LEVEL,_log4cplus_buf.str(),file,line);
    }
}

/**
 * debug - logs DEBUG level messages
 */
void LofarLogger::debug(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::DEBUG_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::DEBUG_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * info - logs INFO level messages
 */
void LofarLogger::info(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::INFO_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::INFO_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * warn - logs WARN level messages
 */
void LofarLogger::warn(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::WARN_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::WARN_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * error - logs ERROR level messages
 */
void LofarLogger::error(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::ERROR_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::ERROR_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * fatal - logs FATAL level messages
 */
void LofarLogger::fatal(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::FATAL_LOG_LEVEL)) 
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::FATAL_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * formatString - creates a formatted string
 */
string LofarLogger::formatString(char* format,...)
{
    va_list args;
    char buffer[1000];
    va_start(args,format);
    vsnprintf(buffer,1000,format,args);
    return string(buffer);
}

/**
 * getLogger - finds a logger in the logger map or creates a new logger
 */
log4cplus::Logger& LofarLogger::getLogger(const string& logger)
{
    TLoggerMap::iterator loggerMapIterator=_loggerMap.find(logger);
    if(loggerMapIterator==_loggerMap.end())
    {
        _loggerMap.insert(TLoggerMap::value_type(logger,log4cplus::Logger::getInstance(logger)));
    }
    loggerMapIterator=_loggerMap.find(logger);
    return (loggerMapIterator->second);
}
