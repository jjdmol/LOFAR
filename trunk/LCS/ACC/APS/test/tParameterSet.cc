//#  tParameterSet.cc: Simple testprogrm to test the ParameterSet class.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #inlcude <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/lofar_iostream.h>
#include <iterator>

#include <APS/ParameterSet.h>
#include <Common/LofarLogger.h>

using namespace std;
using namespace LOFAR;
using namespace LOFAR::ACC::APS;

int main() {
  try {
	INIT_LOGGER("tParameterSet");

	cout << "\nReading in parameterfile 'tParameterSet.in_param'\n";
	cout << ">>>\n";
	ParameterSet		myPS("tParameterSet.in_param");
	cout << "<<<\n";

	ParameterSet		mySecondSet(myPS);

	cout << "\nShowing some values\n";
	cout << "a.b.c=" 			<< myPS.getInt32("a.b.c") << endl;
	cout << "a.b=" 				<< myPS.getInt32("a.b") << endl;
	cout.precision(20);
	cout << "a.b.double="		<< myPS.getDouble("a.b.double") << endl;
	cout << "a.b.lange_naam="	<< myPS.getString("a.b.lange_naam") << endl;

	cout << "\nMerging ParameterSet with file 'tParameterSet.in_merge'\n";
	myPS.adoptFile("tParameterSet.in_merge");

	cout << "\nShowing the same keys again\n";
	cout << "a.b.c=" 			<< myPS.getInt32("a.b.c") << endl;
	cout << "a.b=" 				<< myPS.getInt32("a.b") << endl;
	cout.precision(20);
	cout << "a.b.double="		<< myPS.getDouble("a.b.double") << endl;
	cout << "a.b.lange_naam="	<< myPS.getString("a.b.lange_naam") << endl;

	cout << "a.b.time1="        << myPS.getTime("a.b.time1") << endl;
	cout << "a.b.time2="        << myPS.getTime("a.b.time2") << endl;
	cout << "a.b.time3="        << myPS.getTime("a.b.time3") << endl;

	cout << "\nThe main ParameterSet contains:\n";
	cout << myPS;

	cout << "Fullname of 'b'=" << myPS.locateModule("b") << endl;

	cout << "isValidVersionNr(1.2.3.4)   = " << isValidVersionNr("1.2.3.4") << endl;
	cout << "isValidVersionNr(1.2.3)     = " << isValidVersionNr("1.2.3") << endl;
	cout << "isValidVersionNr(1.2)       = " << isValidVersionNr("1.2") << endl;
	cout << "isValidVersionNr(stable)    = " << isValidVersionNr("stable") << endl;
	cout << "isValidVersionNrRef(1.2.3)  = " << isValidVersionNrRef("1.2.3") << endl;
	cout << "isValidVersionNrRef(1.2)    = " << isValidVersionNrRef("1.2") << endl;
	cout << "isValidVersionNrRef(stable) = " << isValidVersionNrRef("stable") << endl;
	cout << "isValidVersionNrRef(error)  = " << isValidVersionNrRef("error") << endl;
	cout << "isValidVersionNrRef(1.2.3.AndALotOfGarbageBehindTheLastNumberPart)  = " 
		  << isValidVersionNrRef("1.2.3.AndALotOfGarbageBehindTheLastNumberPart");

	ParameterSet		mySubset = myPS.makeSubset("a.b.");
	cout << "\nCreating a subset 'a.b.'\n";
	cout << "\nSubset a.b. contains:\n";
	cout << mySubset;

	cout << "\nTrying to read a non-existing key\n"; 
	cout << ">>>\n";
	try {
		myPS.getInt32("is.er.niet");
	}
	catch (LOFAR::Exception& ex) {
	  cout << "<<<\n";
	  cout << "Told you the key didn't exist." << endl;
	}

	cout << "\nFinally write the parameterset to 'newset.stdout'\n";
	myPS.writeFile("tParameterSet_tmp.newset.stdout");

	try {
		cout << "\ntesting getInt32Vector\n";
		vector<int32> intVector = myPS.getInt32Vector("vtest.intVector1Dim");
		cout << intVector.size() << " elements in intVector1Dim\n";
		copy (intVector.begin(), intVector.end(), 
		      std::ostream_iterator<int, char>(cout, ","));
		cout << endl;

		cout << "trying to read single int as vector\n";
		intVector = myPS.getInt32Vector("a.b.c");
		cout << intVector.size() << " elements in a.b.c\n";
		copy (intVector.begin(), intVector.end(), 
		      std::ostream_iterator<int, char>(cout, ","));
		cout << endl;
	}
	catch (LOFAR::Exception& ex) {
		LOG_DEBUG_STR ("Exception:" << ex.what());
	}

	// Iterate through all keys.
	cout << endl << "Iterate over all keys ..." << endl;
	for (ParameterSet::iterator iter = myPS.begin();
	     iter != myPS.end();
	     iter++) {
	  cout << iter->first << endl;
	}

  } catch (std::exception& ex) {
    cout << "Unexpected exception: " << ex.what() << endl;
    return 1;
  }
  return 0;

}
