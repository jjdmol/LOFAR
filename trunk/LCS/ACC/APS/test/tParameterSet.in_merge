# comment line
a.b.d=7
# the next two lines overlap with the KVs in test.ps
a.b.lange_naam="dit is nu een andere naam geworden met extra spaties aan het einde  "
a.b.c=5

e.g=een voorbeeld
egg=een ei
