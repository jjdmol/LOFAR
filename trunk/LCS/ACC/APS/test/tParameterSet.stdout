
Reading in parameterfile 'tParameterSet.in_param'
>>>
WARN - Key a.b.c is defined twice. Ignoring first value.
<<<

Showing some values
a.b.c=700
a.b=0
a.b.double=3.1415926535897931
a.b.lange_naam=dit is een wat langere naam

Merging ParameterSet with file 'tParameterSet.in_merge'

Showing the same keys again
a.b.c=5
a.b=0
a.b.double=3.1415926535897931
a.b.lange_naam=dit is nu een andere naam geworden met extra spaties aan het einde  
a.b.time1=15
a.b.time2=900
a.b.time3=54000

The main ParameterSet contains:
[a.b],[0..3]
[a.b.c],[5]
[a.b.c.d.e.f.g.h],[tjee wat een levels!]
[a.b.d],[7]
[a.b.double],[3.1415926535897932384626433832795]
[a.b.lange_naam],[dit is nu een andere naam geworden met extra spaties aan het einde  ]
[a.b.time1],[15 s]
[a.b.time2],[15m]
[a.b.time3],[15 hour]
[a.c.a],[nog een afkorting]
[e.g],[een voorbeeld]
[egg],[een ei]
[vtest.intVector1Dim],[[ 3, "20485", 	002, '4904' , 25 ]]
Fullname of 'lange_naam=a.b.lange_naam
isValidVersionNr(1.2.3.4)   = 0
isValidVersionNr(1.2.3)     = 1
isValidVersionNr(1.2)       = 0
isValidVersionNr(stable)    = 0
isValidVersionNrRef(1.2.3)  = 1
isValidVersionNrRef(1.2)    = 0
isValidVersionNrRef(stable) = 1
isValidVersionNrRef(error)  = 0
isValidVersionNrRef(1.2.3.AndALotOfGarbageBehindTheLastNumberPart)  = 0
Creating a subset 'a.b.'

Subset a.b. contains:
[c],[5]
[c.d.e.f.g.h],[tjee wat een levels!]
[d],[7]
[double],[3.1415926535897932384626433832795]
[lange_naam],[dit is nu een andere naam geworden met extra spaties aan het einde  ]
[time1],[15 s]
[time2],[15m]
[time3],[15 hour]

Trying to read a non-existing key
>>>
DEBUG - Key is.er.niet unknown
<<<
Told you the key didn't exist.

Finally write the parameterset to 'newset.stdout'

testing getInt32Vector
5 elements in intVector1Dim
3,20485,2,4904,25,

Iterate over all keys ...
a.b
a.b.c
a.b.c.d.e.f.g.h
a.b.d
a.b.double
a.b.lange_naam
a.b.time1
a.b.time2
a.b.time3
a.c.a
e.g
egg
vtest.intVector1Dim
